### ISSUE TITLE
[ArrayAndObjectConstructors] Description enhancement

### ISSUE THREAD

**Issue by AMorgaut** [SonarSource/SonarJS#890]
**I want to provide the following feedback.**

**Rule key:** ArrayAndObjectConstructors

Many developers really dislike this rule as they think they should still be able to use "new Array(size)" which looks handy to create an array with a predefined length

Some linter allow to ignore this pattern, but from my experience it is still dangerous

Find below a proposal to extend the rule documentation and prevent developers from just adding plenty of NOSONAR comments:

## Array Size Use Case

This form is often still used by developers who consider it "safe"
It is unfortunately not.... Mistyped value can very easily create very wrong arrays

```javascript
var a = new Array(5); // Creates an empty array with a size of 5  without defined elements
var a = new Array("5"); // Creates an array with 1 element initialized to "5"
```

Safest solutions are

In ES5:

```javascript
var a = []; // Creates an empty array 
a.length = size; // Set the size of the array
```

In ES6:

The [Array.from()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from) static method is very handy to initialize almost any kind of array

```javascript
const a1 = Array.from({length: n}); // Creates an array with a size of 'n'
const a2 = Array.from({length: n}, (value, index) => index); // Creates an array with 'n' defined elements
```

cf: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from#Using_arrow_functions_and_Array.from

## Sparse Array Warning

Note that setting the size of an Array without initializing its values lead to sparse array situation

JavaScript do have sparse Array optimization mechanism that can lead to unexpected situation 

```javascript
const a = new Array(5); // Creates an array with the size 5
const b = a.map(callback); // callback is never called
```

Array methods like `filter()`, `forEach()`,  `every()`,  `some()`, `map()`, `reduce()`,  and `reduceRight()`, will only iterate over "each element present in the array" which means: "only initialized elements", and ignore all other ones (Note: elements can be present  / initialized with `undefined` value, the can be removed with `delete`).

**Comment by AMorgaut** [comment]
Note:
I already included it on our server using the awsome "Extend Description" feature, but I thought it would worth to be natively in the rule description
